# GEO-RNAseq-Progress
Daily Updatesï¼šRecording of project progress

## ä»»åŠ¡èƒŒæ™¯ï¼š
åŸºäºæŒ‡å®šçš„GEOç¼–å·GSE255647ï¼Œé€šè¿‡RNA-seqæ•°æ®çš„ä¸‹è½½ã€é¢„å¤„ç†ã€æ¯”å¯¹ã€å®šé‡ã€å·®å¼‚è¡¨è¾¾åˆ†æå’ŒåŠŸèƒ½å¯Œé›†åˆ†æï¼Œå®Œæˆè½¬å½•ç»„æ•°æ®çš„å…¨æµç¨‹åˆ†æï¼Œå¹¶ä»¥Markdownå½¢å¼è®°å½•è¿‡ç¨‹ï¼Œæœ€ç»ˆä»¥GitHubä»“åº“å±•ç¤ºã€‚

**ä»»åŠ¡è¦æ±‚**ï¼š åˆ†æSARS-CoV-2ä»¥1 MOIæ„ŸæŸ“Calu-3/2B4ç»†èƒç³»12håï¼Œè½¬å½•ç»„æ°´å¹³çš„å˜åŒ–ã€‚

# âœ”å·²å®Œæˆ
## æ¨¡å—1ï¼šä»»åŠ¡å‡†å¤‡

**å®ç”¨æ“ä½œ**

lsï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šlist filesï¼‰: åˆ—å‡ºç›®å½•åŠæ–‡ä»¶å

cdï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šchange directoryï¼‰ï¼šåˆ‡æ¢ç›®å½•

pwdï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šprint work directoryï¼‰ï¼šæ˜¾ç¤ºç›®å‰çš„ç›®å½•

mkdirï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šmake directoryï¼‰ï¼šåˆ›å»ºä¸€ä¸ªæ–°çš„ç›®å½•

rmdirï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šremove directoryï¼‰ï¼šåˆ é™¤ä¸€ä¸ªç©ºçš„ç›®å½•

cpï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šcopy fileï¼‰: å¤åˆ¶æ–‡ä»¶æˆ–ç›®å½•

rmï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šremoveï¼‰: åˆ é™¤æ–‡ä»¶æˆ–ç›®å½•

mvï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šmove fileï¼‰: ç§»åŠ¨æ–‡ä»¶ä¸ç›®å½•ï¼Œæˆ–ä¿®æ”¹æ–‡ä»¶ä¸ç›®å½•çš„åç§°



**å®Œæˆ**ï¼š

åŸºäº`micromamba`é…ç½®ç¯å¢ƒï¼Œå®‰è£…å¹¶ç†Ÿæ‚‰ä»¥ä¸‹å·¥å…·ï¼š

1. fastqcã€multiqcã€fastpï¼šè´¨æ§ä¸æ¸…æ´—

2. hisat2ã€stringtieï¼šæ¯”å¯¹ä¸å®šé‡

3. samtoolsï¼šBAMæ–‡ä»¶æ“ä½œ

4. Rä¸DESeq2ã€clusterProfilerã€ggplot2

5. æ³¨å†ŒGitHubè´¦å·ï¼ŒæŒæ¡åŸºæœ¬æ“ä½œ

**æ€è€ƒé¢˜**ï¼š

1. å¦‚ä½•åœ¨æ²¡æœ‰ç®¡ç†å‘˜æƒé™ä¸‹å®‰è£…fastqcï¼Ÿ

   ç­”ï¼šæœ‰å¤šç§æ–¹æ³•ï¼ŒåŒ…æ‹¬micromambaã€æ‰‹åŠ¨ä¸‹è½½äºŒè¿›åˆ¶æ–‡ä»¶ä»¥åŠåœ¨Javaç¯å¢ƒä¸‹é€šè¿‡æºç ç¼–è¯‘ç­‰æ–¹å¼ã€‚è¿™é‡Œä½¿ç”¨micromambaã€biocondaï¼ˆæˆ–é•œåƒï¼‰ï¼š
```
   micromamba create -p ~/micromamba_envs/fastqc_env -c bioconda -c conda-forge fastqc
```
2. GitHubçš„READMEæ–‡ä»¶æœ‰ä»€ä¹ˆä½œç”¨ï¼Ÿ
   
   é¡¹ç›®çš„â€œé—¨é¢â€å’Œâ€œä½¿ç”¨è¯´æ˜ä¹¦â€ï¼Œåƒè¿™æ ·ğŸ‘€ã€‚


## æ¨¡å—2ï¼šæ•°æ®ä¸‹è½½ä¸é¢„å¤„ç†

**ç›®æ ‡**ï¼šä»GEOä¸‹è½½åŸå§‹æ•°æ®ï¼Œè¿›è¡Œè´¨æ§å’Œæ¸…æ´—ã€‚

### ä»»åŠ¡ï¼š

**1. ä½¿ç”¨wgetã€curlæˆ–asperaä¸‹è½½æŒ‡å®šGEOç¼–å·çš„FASTQæ•°æ®(ä»¥SRR27961778ä¸ºä¾‹ï¼‰**
```
#ä¸‹è½½SRAæ–‡ä»¶
    wget -O SRR27961778.sra \"https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR27961778/SRR27961778"
#è½¬æ¢ä¸ºFASTQæ–‡ä»¶
   fastq-dump --split-files SRR27961778.sra
#å¯ä»¥å­˜ä¸ºå‹ç¼©åŒ…
   gzip SRR27961779_1.fastq
```

**2. è´¨æ§ï¼šfastqcå’Œmultiq**
```
#ç”ŸæˆåŸå§‹æ•°æ®è´¨æ§æŠ¥å‘Š
   fastqc SRR27961778_1.fastq.gz SRR27961778_2.fastq.gz -o ./raw_fastqc_results 
   cd /mnt/alamo01/users/chenyun730/program/test/raw_fastqc_results
   multiqc ./SRR*
```

**3. æ¸…æ´—ï¼šfastpï¼ˆå»ä½è´¨ã€å»æ¥å¤´ï¼‰**
```
#å…ˆåœ¨testé‡Œæ–°å»ºä¸€ä¸ªæ–‡ä»¶å¤¹cleanâ€”â€”dataç”¨äºå­˜å‚¨æ¸…æ´—åçš„æ•°æ®
$ cd /mnt/alamo01/users/chenyun730/program/test
$ mkdir clean_data
$ for F in /mnt/alamo01/users/chenyun730/program/test/raw_data/*_1.fastq.gz; do    
 R=${F%_*}_2.fastq.gz;
 BASE=${F##*/}; SAMPLE=${BASE%_*};
 time fastp -i $F -I $R
   -o /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_1.fp.gz
   -O /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_2.fp.gz
   -e 25 -q 30 -u 10 -r -M 30 -W 5 -w 64 -f 10 -F 10
   --adapter_sequence AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC
   --adapter_sequence_r2 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT
   -h /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_report.html
   -j /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_report.json &
done
#e â€”â€”å¹³å‡è´¨é‡é˜ˆå€¼ï¼› q â€”â€”è´¨é‡é˜ˆå€¼ï¼ˆé»˜è®¤ä¸º15ï¼Œè¿™é‡Œæé«˜åˆ°30ä¸ºæ›´ä¸¥æ ¼ï¼‰ï¼›
u â€”â€”å…è®¸è¾¾æ ‡ç¢±åŸºçš„æ¯”ä¾‹ä¸º10%ï¼› r â€”â€”å¯ç”¨ polyG ä¿®å‰ªï¼ˆé€‚ç”¨äº NovaSeqï¼‰ï¼›
M â€”â€”polyG çš„æœ€å°é•¿åº¦é˜ˆå€¼ï¼›
W 5 â€”â€”æ»‘åŠ¨çª—å£å¤§å°ï¼ˆ5bpï¼‰ï¼› w 64 â€”â€”ä½¿ç”¨64ä¸ªçº¿ç¨‹ï¼›
f 10 F 10 â€”â€”åˆ‡é™¤ Read 1/2 å‰ 10 bp
```

**4. è¾“å‡ºï¼šè´¨æ§æŠ¥å‘Šã€æ¸…æ´—åæ•°æ®**
```
multiqc ./ -n "Cleaned_Data_QC"
```

**æ€è€ƒé¢˜**ï¼š

å¦‚æœè´¨æ§æŠ¥å‘Šä¸­ç¢±åŸºè´¨é‡åä½ï¼Œå¦‚ä½•è°ƒæ•´fastpçš„å‚æ•°ï¼Ÿ
 
1.å…¨å±€è°ƒæ•´ï¼š æé«˜è´¨é‡é˜ˆå€¼qã€é™ä½å…è®¸æœªè¾¾æ ‡ç¢±åŸºçš„æ¯”ä¾‹uã€æé«˜å¹³å‡è´¨é‡é˜ˆå€¼eï¼›
   
2.è‹¥5' æˆ– 3' ç«¯è´¨é‡å·®ï¼Œåºåˆ—å¼€å¤´æˆ–æœ«å°¾è´¨é‡æ˜æ˜¾ä¸‹é™ï¼šæé«˜é¦–ä½åˆ‡é™¤çš„bpæ•°ï¼ˆ-fã€-Fã€-tï¼‰ï¼›
   
3.è‹¥æ¥å¤´åºåˆ—æ®‹ç•™ï¼ˆMultiQC çš„ Adapter Content éƒ¨åˆ†æ˜¾ç¤ºè¶…æ ‡ï¼‰ï¼šè‡ªåŠ¨æ£€æµ‹æ¥å¤´ï¼ˆå³ä½¿å·²æŒ‡å®šåºåˆ—ï¼‰æˆ–è€…é¢å¤–åˆ‡é™¤5bpï¼ˆ-trim_front1 5);
   
4.è‹¥é«˜æ¯”ä¾‹çš„é‡å¤åºåˆ—æˆ–ä½å¤æ‚åº¦åºåˆ—ï¼šè¿‡æ»¤æ‰é•¿åº¦< 50bp çš„ readsï¼ˆé»˜è®¤ 15ï¼‰-l 50ï¼›ç§»é™¤ä½å¤æ‚åº¦åºåˆ—-low_complexity_filterï¼›
   
5.è‹¥å±€éƒ¨åŒºåŸŸè´¨é‡æ³¢åŠ¨å¤§ï¼šå‡å°æ»‘åŠ¨çª—å£å¤§å°-Wï¼›æ»‘åŠ¨çª—å£å¹³å‡è´¨é‡é˜ˆå€¼ä» 30 é™è‡³ 25ï¼ˆæ›´å®½æ¾ï¼‰-M
    

## æ¨¡å—3ï¼šæ¯”å¯¹ä¸å®šé‡

**ç›®æ ‡**ï¼šåŸºäºå‚è€ƒåŸºå› ç»„æ„å»ºç´¢å¼•å¹¶è¿›è¡Œæ¯”å¯¹ä¸å®šé‡ã€‚

### ä»»åŠ¡ï¼š

ä¸‹è½½ç‰©ç§å‚è€ƒåŸºå› ç»„å’ŒåŸºå› æ³¨é‡Šï¼ˆGTFæ–‡ä»¶ï¼‰
```
 #ä¸‹è½½å‚è€ƒåŸºå› ç»„çš„FASTQå’ŒGTFæ–‡ä»¶
 cd /mnt/alamo01/users/chenyun730/program/test/compare
 mkdir homo_sapiens
 wget ftp://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
 wget ftp://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh38.104.gtf.gz
 # è§£å‹æ–‡ä»¶
 gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
 gunzip Homo_sapiens.GRCh38.104.gtf.gz
```

ç”¨hisat2-buildæ„å»ºç´¢å¼•ï¼Œhisat2æ¯”å¯¹
```
# å‚è€ƒåŸºå› ç»„ GRCh38.fa æ„å»ºç´¢å¼•ï¼ˆæäº¤è„šæœ¬å¹¶æŠ•é€’ä»»åŠ¡ï¼‰
mkdir -p /mnt/alamo01/users/chenyun730/program/test/scripts
mkdir -p /mnt/alamo01/users/chenyun730/program/test/logs

#æ–°å»ºshellè„šæœ¬
vim hisat2_index.sh
#! /usr/bin/sh
#micromamba activate R441
cd /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data
hisat2-build -p 64 GRCh38.fa GRCh38
2> /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/home_data

#èµ‹äºˆè„šæœ¬æ‰§è¡Œæƒé™
chmod 777 hisat2_index.sh

#æäº¤ä»»åŠ¡(æ— éœ€æŒ‡å®šé˜Ÿåˆ—ï¼‰
qsub -cwd -V -l cpu=64:mem=64G -q fast -N hisat2_index /mnt/alamo01/users/chenyun730/program/test/scripts/hisat2_index.sh
# qstatæŸ¥çœ‹ä»»åŠ¡é˜Ÿåˆ—ï¼›qstat -f ID æŸ¥çœ‹ä»»åŠ¡ï¼› qdel åˆ é™¤ä»»åŠ¡

# æ£€æŸ¥ç´¢å¼•æ–‡ä»¶
ls -lh /mnt/alamo01/users/chenyun730/program/test/compare/homo_sapiens/genome_index/

# è¿›è¡Œæ¯”å¯¹
#! /bin/bash
#source /mnt/alamo01/users/chenyun730/bin/micromamba
#micromamba activate R441
cd /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/
hisat2 -k 1 -p 64 \
  -x /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/GRCh38 \
  -S SRR27961779.sam \
  --novel-splicesite-outfile SRR27961779_junction.bed \
  --no-unal --dta \
  --un-conc-gz SRR27961779_unmapped.fq.gz \
  -1 /mnt/alamo01/users/chenyun730/program/test/clean_data/SRR27961779_cleaned_1.fp.gz \
  -2 /mnt/alamo01/users/chenyun730/program/test/clean_data/SRR27961779_cleaned_2.fp.gz \
  2> SRR27961779.align.stats

#å¤šä¸ªæ–‡ä»¶è¿›è¡Œæ¯”å¯¹
#! /bin/bash
#source /mnt/alamo01/users/chenyun730/bin/micromamba
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
cd /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/
for SAMPLE in "${SAMPLES[@]}"; do
hisat2 -k 1 -p 64 \
  -x /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/GRCh38 \
  -S ${SAMPLE}.sam \
  --novel-splicesite-outfile ${SAMPLE}_junction.bed \
  --no-unal --dta \
  --un-conc-gz ${SAMPLE}_unmapped.fq.gz \
  -1 /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_1.fp.gz \
  -2 /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_2.fp.gz \
  2> ${SAMPLE}.align.stats
echo "Submitted alignment job for ${SAMPLE}"
done
```
ä½¿ç”¨samtoolså’Œstringtieè¿›è¡Œè½¬å½•æœ¬ç»„è£…å’ŒåŸºå› å®šé‡
```
#è½¬æ¢SAMä¸ºBAMå¹¶æ’åº
$ vim samtools.sh
#! /bin/bash
#source /mnt/alamo01/users/chenyun730/bin/micromamba
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
echo "Processing ${SAMPLE}..."
  mv /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/${SAMPLE}.sam /mnt/alamo01/users/chenyun730/program/test/alignment/${SAMPLE}
  cd /mnt/alamo01/users/chenyun730/program/test/alignment/${SAMPLE}
  samtools sort -@ 8 -o ${SAMPLE}_sorted.bam ${SAMPLE}.sam
  samtools index ${SAMPLE}_sorted.bam
  echo "Finished processing ${SAMPLE}"
done


#ä½¿ç”¨stringtieè¿›è¡Œè½¬å½•æœ¬ç»„è£…å’ŒåŸºå› å®šé‡
$ vim stringtie.sh
#! /bin/bash
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
        echo "Processing ${SAMPLE}..."
 cd /mnt/alamo01/users/chenyun730/program/test/alignment/${SAMPLE}
        stringtie /mnt/alamo01/users/chenyun730/program/test/alignment/${SAMPLE}/${SAMPLE}_sorted.bam \
  -G /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/Homo_sapiens.GRCh38.109.gtf \
  -o /mnt/alamo01/users/chenyun730/program/test/alignment/${SAMPLE}.gtf \
  -p 64 \
  -B
        echo "Finished processing ${SAMPLE}"
done

find /mnt/alamo01/users/chenyun730/program/test/alignment/ -type f -name "*.gtf" > gtf_list.txt
 stringtie --merge -G /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/Homo_sapiens.GRCh38.109.gtf
-o merged.gtf   gtf_list.txt
 
 #ç”¨sh.æäº¤ï¼Œè®¡ç®—æ ·æœ¬çš„åŸºå› è¡¨è¾¾é‡ï¼ˆä»¥åˆå¹¶åçš„åŸºå› ç»„åšå‚è€ƒï¼‰
#! /bin/bash
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
    echo "Processing ${SAMPLE}..."
    stringtie /mnt/alamo01/users/chenyun730/program/test/alignment/${SAMPLE}/${SAMPLE}_sorted.bam \
        -G /mnt/alamo01/users/chenyun730/program/test/alignment/merged.gtf \
         -o /mnt/alamo01/users/chenyun730/program/test/quantify/${SAMPLE}.gtf \
         -e \
         -B \
         -A /mnt/alamo01/users/chenyun730/program/test/quantify/${SAMPLE}/${SAMPLE}_gene_abundance.tab
     echo "Finished processing ${SAMPLE}"
done
```

æå–FPKMåˆ°ä¸€ä¸ªçŸ©é˜µ
```
vim fpkm.sh
#! /bin/bash
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
    awk 'BEGIN {OFS="\t"} {if (NR>1) print $9, $6, $12}' /mnt/alamo01/users/chenyun730/program/test/quantify/${SAMPLE}/${SAMPLE}/t_data.ctab > /mnt/alamo01/users/chenyun730/program/test/quantify/${SAMPLE}/${SAMPLE}_FPKM.txt
done
OUTPUT_FILE="/mnt/alamo01/users/chenyun730/program/test/quantify/all_samples_FPKM.txt"
echo -n "gene_id" > $OUTPUT_FILE
for SAMPLE in "${SAMPLES[@]}"; do
    echo -n -e "\tFPKM_${SAMPLE}" >> $OUTPUT_FILE
done
echo "" >> $OUTPUT_FILE
FIRST_SAMPLE="${SAMPLES[0]}"
FIRST_FILE="/mnt/alamo01/users/chenyun730/program/test/quantify/${FIRST_SAMPLE}/${FIRST_SAMPLE}_FPKM.txt"
cut -f1,2 "$FIRST_FILE" | tail -n +2 > /tmp/combined.tmp
for ((i=1; i<${#SAMPLES[@]}; i++)); do
    SAMPLE="${SAMPLES[$i]}"
    SAMPLE_FILE="/mnt/alamo01/users/chenyun730/program/test/quantify/${SAMPLE}/${SAMPLE}_FPKM.txt"
    cut -f2 "$SAMPLE_FILE" | tail -n +2 > "/tmp/${SAMPLE}_col.tmp"
    paste /tmp/combined.tmp "/tmp/${SAMPLE}_col.tmp" > /tmp/combined_new.tmp
    mv /tmp/combined_new.tmp /tmp/combined.tmp
done
cat /tmp/combined.tmp >> "$OUTPUT_FILE"
rm /tmp/*_col.tmp /tmp/combined.tmp
echo "FPKM data for all samples has been merged into $OUTPUT_FILE"

#ä¸‹è½½prepDE.py3è¿è¡Œå¾—åˆ° .csvæ–‡ä»¶ï¼ˆä½¿ç”¨countå¯è¿›è¡Œï¼‰è¿™é‡Œä¸ç”¨
python3 /mnt/alamo01/users/chenyun730/program/test/scripts/prepDE.py3 -i /mnt/alamo01/users/chenyun730/program/test/quantify/quantify_gtf_list.txt -v
python3 /mnt/alamo01/users/chenyun730/program/test/scripts/prepDE.py3 -i .  -v  #æœ€ä½³ç‰ˆæœ¬

```

è¾“å‡ºï¼šæ¯”å¯¹ç»“æœï¼ˆBAMï¼‰ã€å®šé‡çŸ©é˜µ


**æ€è€ƒé¢˜**ï¼š

ä¸ºä»€ä¹ˆéœ€è¦æ„å»ºå‚è€ƒåŸºå› ç»„ç´¢å¼•ï¼Ÿ

1. åŠ é€Ÿå¯¹æ¯”ï¼Œå‡å°‘å¯¹æ¯”æ—¶é—´ï¼›
   
2. å ç”¨æ›´å°‘çš„å†…å­˜ï¼›
   
3. æ”¯æŒå¤æ‚æ¯”å¯¹æ¨¡å¼ï¼š
     ï¼ˆaï¼‰å‰ªæ¥æ¯”å¯¹ï¼šRNA-seqæ•°æ®éœ€æ£€æµ‹å¤–æ˜¾å­è¿æ¥ï¼Œç´¢å¼•ä¼šé¢„å­˜å‰ªåˆ‡ä½ç‚¹ä¿¡æ¯ï¼ˆHISAT2çš„snp_tranç´¢å¼•ï¼‰ï¼›
     ï¼ˆbï¼‰çªå˜å®¹å¿ï¼šåŒ…å«SNP/çªå˜çš„ç´¢å¼•ï¼ˆå¦‚genome_snp_tranï¼‰å¯æé«˜å¤šæ€æ€§æ ·æœ¬çš„æ¯”å¯¹ç‡ã€‚

### æ¨¡å—4ï¼šæ•°æ®åˆ†æä¸å¯è§†åŒ–

**ç›®æ ‡**ï¼šå¯¹å®šé‡æ•°æ®è¿›è¡Œè´¨æ§å¹¶ä¸GEOåŸä½œè€…çš„è¡¨è¾¾çŸ©é˜µæ¯”è¾ƒã€‚

**ä»»åŠ¡**ï¼š

1. è¡¨è¾¾çŸ©é˜µçš„log2è½¬æ¢å’Œæ ‡å‡†åŒ–
```
#åŠ è½½RåŒ…å’Œ.csvæ•°æ®
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(FactoMineR)
library(factoextra)
fpkm <- read.table("/mnt/alamo01/users/chenyun730/program/test/quantify/all_samples_FPKM.txt", header=TRUE, sep="\t", row.names=1, check.names=FALSE) #è¿è¡Œåç”±äºæœ‰é‡å¤çš„gene_idè€ŒæŠ¥é”™
fpkm <- read.table("/mnt/alamo01/users/chenyun730/program/test/quantify/all_samples_FPKM.txt",header=TRUE, sep="\t", check.names=FALSE)
library(dplyr)#å–å‡å€¼åˆå¹¶
sum(duplicated(fpkm[, 1]))
fpkm_unique <- fpkm %>%
group_by(gene_id = fpkm[,1]) %>%
summarise(across(.cols = everything(), .fns = mean))
fpkm_matrix <- as.data.frame(fpkm_unique)
rownames(fpkm_matrix) <- fpkm_matrix$gene_id
fpkm_matrix$gene_id <- NULL

#å†æ£€æŸ¥ä¸€ä¸‹
 sum(duplicated(rownames(fpkm_matrix))) #æ­¤æ—¶è¾“å‡ºä¸º0ï¼Œè¯´æ˜å·²ç»æ²¡æœ‰é‡å¤çš„gene_id

#å–log2å¯¹æ•°
fpkm_log2 <- log2(fpkm_matrix + 1)
```

2. çƒ­å›¾
```
/mnt/alamo01/users/chenyun730/program/test/results/
pheatmap(fpkm_log2[top500, ],
         scale="row",
         show_rownames=FALSE,
         cluster_rows=TRUE,
         cluster_cols=TRUE,
         main="Top 500 Most Variable Genes")
dev.off()

# ä¿®æ”¹æ–‡ä»¶å
expr <- read.csv("your_expression_matrix.csv", row.names = 1)  # or .tsv with read.delim   # è¯»å–è¡¨è¾¾çŸ©é˜µ
srr_ids <- c("SRR27961778", "SRR27961779", "SRR27961780", "SRR27961787", "SRR27961788", "SRR27961789") #æŒ‡å®šä½ æƒ³æå–çš„æ ·æœ¬
new_names <- c("Mock-1", "Mock-2", "Mock-3", "SARS-2-1", "SARS-2-2", "SARS-2-3") #å¯¹åº”çš„ç›®æ ‡å‘½å
expr_subset <- expr[, srr_ids] # æå–æ„Ÿå…´è¶£çš„6ä¸ªæ ·æœ¬
colnames(expr_subset) <- new_names  #é‡å‘½ååˆ—
write.csv(expr_subset, file = "subset_6samples_renamed.csv") #ä¿å­˜ç»“æœ

```

3. PCAå›¾
```
fpkm_log2_filtered <- fpkm_log2[apply(fpkm_log2, 1, function(x) sd(x) != 0), ] #è¿‡æ»¤æ‰NA
cat("Retained genes for PCA:", nrow(fpkm_log2_filtered), "\n")
fpkm_log2[is.na(fpkm_log2)] <- 0
pca_input <- t(fpkm_log2)
pca <- prcomp(pca_input, scale.=TRUE)
pca_df <- data.frame(pca$x)
pca_df$Sample <- rownames(pca_df)
library(ggplot2)
pdf("PCA_plot.pdf", width=8, height=6)
ggplot(pca_df, aes(PC1, PC2, label=Sample)) +
  geom_point(size=4, color="steelblue") +
  geom_text(vjust=-1) +
  theme_minimal() +
  labs(title="PCA of Samples", x="PC1", y="PC2")
dev.off()
```
ä¸‹è½½åŸä½œè€…çš„çŸ©é˜µè¿›è¡Œæ¯”å¯¹

1.è‡ªå·±çš„countsçŸ©é˜µå…ˆç»˜å›¾æ£€æŸ¥å¼‚è´¨æ€§
```
micromamba install -c conda-forge -c bioconda bioconductor-geoquery
# æ£€æŸ¥é‡å¤çš„ gene_id
sum(duplicated(rownames(counts)))  # è¿”å›é‡å¤åŸºå› æ•°é‡ï¼Œè¿”å›0
sum(is.na(counts)))
counts[is.na(counts)] <- 0# æ›¿æ¢ NA ä¸º 0ï¼ˆæˆ–æ ¹æ®éœ€æ±‚è¿‡æ»¤ä½è¡¨è¾¾åŸºå› ï¼‰
keep <- rowSums(counts > 1) >= 3
counts <- counts[keep, ]# ä¿ç•™è‡³å°‘åœ¨ 3 ä¸ªæ ·æœ¬ä¸­ count > 1 çš„åŸºå› 
colData <- data.frame(
  group = factor(rep(c("Mock", "SARS-2"), each = 3)),
  row.names = colnames(counts)
) # å®šä¹‰æ ·æœ¬åˆ†ç»„ï¼ˆMock vs SARS-2ï¼‰

# æ„å»º DESeqDataSet
dds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = colData,
  design = ~ group
)
# è½¬æ¢æ•°æ®ï¼ˆblind=TRUE é€‚ç”¨äºæ— ç›‘ç£åˆ†æï¼‰
rld <- rlog(dds, blind = TRUE)  #rlog æ›´é€‚åˆå°æ•°æ®é›†ï¼ˆæ ·æœ¬æ•° < 30ï¼‰
rld_matrix <- assay(rld)  # æå–è½¬æ¢åçš„çŸ©é˜µ
#ä¿®æ”¹å‘½åé‡Œçš„â€œ-â€
colData$group <- gsub("-", "_", colData$group)
ann_colors <- list(Group = c(Mock = "blue", SARS_2 = "red")) #çƒ­å›¾çš„æ­¥éª¤
#ç»˜åˆ¶ PCA å›¾ï¼ˆåˆ†ç»„ç€è‰²ï¼‰
pca_data <- plotPCA(rld, intgroup = "group", returnData = TRUE)
percent_var <- round(100 * attr(pca_data, "percentVar"))

morandi_blue <- "#89CFF0" 
morandi_orange <- "#FFB347"
 ggplot(pca_data, aes(PC1, PC2, color = group)) +
  geom_point(size = 3, alpha = 0.8) +
  scale_color_manual(values = c("Mock" = "#89CFF0", "SARS-2" = "#FFB347")) +
  xlab(paste0("PC1 (", percent_var[1], "%)")) +
  ylab(paste0("PC2 (", percent_var[2], "%)")) +
  ggtitle("PCA Plot of my_matrix") +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
    legend.position = "right"
  ) +
  geom_text(aes(label = name), vjust = 1.5, size = 4)
ggsave("/mnt/alamo01/users/chenyun730/program/test/results/PCA_plot.png", width = 8, height = 6, dpi = 300)
#ç»˜åˆ¶çƒ­å›¾
library(pheatmap)
top_genes <- head(order(rowVars(rld_matrix), decreasing = TRUE), 500)
annotation_col <- data.frame(
  Group = colData$group,
  row.names = colnames(counts))
ann_colors <- list(
  Group = c(Mock = "#89CFF0", SARS_2 = "#FFB347"))
pdf("/mnt/alamo01/users/chenyun730/program/test/results/heatmap_plot.pdf",width = 10, height = 12, pointsize = 15 )
pheatmap(
  rld_matrix[top_genes, ],
  scale = "row",
  cluster_rows = TRUE,
  cluster_cols = TRUE,
  show_rownames = FALSE,
  annotation_col = annotation_col,
  annotation_colors = ann_colors,
  main = "Top 500 Genes by Variance (rlog-transformed)",
  color = colorRampPalette(c("navy", "white", "firebrick3"))(100)
)
dev.off()
```
2.ä¸‹è½½GEOåŸä½œè€…çš„çŸ©é˜µè¿›è¡Œåˆ†æï¼ˆGSE255647_all.counts.tsv.gzï¼‰
```
# è¯»å– count çŸ©é˜µï¼ˆgz æ ¼å¼çš„ TSV æ–‡ä»¶ï¼‰
counts <- read.table("GSE255647_all.counts.tsv.gz", 
                     header = TRUE, 
                     sep = "\t", 
                     row.names = 1, 
                     check.names = FALSE)
head(counts[, 1:6])
str(counts)
# æå–ç›®æ ‡æ ·æœ¬
selected_samples <- c("SRR27961778", "SRR27961779", "SRR27961780", 
                      "SRR27961787", "SRR27961788", "SRR27961789")

counts_subset <- counts[, selected_samples]

# é‡å‘½åä¸ºæ›´æ˜“æ‡‚çš„æ ¼å¼
colnames(counts_subset) <- c("Mock-1", "Mock-2", "Mock-3", 
                             "SARS-2-1", "SARS-2-2", "SARS-2-3")
# é€‰æ‹©ç›®æ ‡æ ·æœ¬
target_samples <- c("MOCK_12hpi_rep1", "MOCK_12hpi_rep2", "MOCK_12hpi_rep3",
                    "SARS-2_12hpi_rep1", "SARS-2_12hpi_rep2", "SARS-2_12hpi_rep3")

counts_subset <- counts[, target_samples]

# é‡å‘½ååˆ—
colnames(counts_subset) <- c("Mock-1", "Mock-2", "Mock-3",
                             "SARS-2-1", "SARS-2-2", "SARS-2-3")
library(DESeq2)

# æ„å»ºæ ·æœ¬ä¿¡æ¯
group <- factor(c("Mock", "Mock", "Mock", "SARS-2", "SARS-2", "SARS-2"))
col_data <- data.frame(row.names = colnames(counts_subset),
                       condition = group)

# æ„å»º DESeq2 å¯¹è±¡
dds <- DESeqDataSetFromMatrix(countData = counts_subset,
                              colData = col_data,
                              design = ~ condition)

# è¿‡æ»¤ä½è¡¨è¾¾åŸºå› ï¼ˆæ¨èï¼‰
dds <- dds[rowSums(counts(dds)) > 10, ]

# è¿è¡Œæ ‡å‡† DESeq2 æµç¨‹
dds <- DESeq(dds)

# ç»˜åˆ¶çƒ­å›¾+PCAå›¾
rld <- rlog(dds, blind = TRUE) #å¦ä¸€ä¸ª vsd <- vst(dds, blind = TRUE)
png("/mnt/alamo01/users/chenyun730/program/test/results/PCA_plot.png", 
    width = 800, height = 800)
plotPCA(rld, intgroup = "condition")
dev.off()

library(pheatmap)
library(matrixStats)
top_var_genes <- head(order(rowVars(assay(rld)), decreasing = TRUE), 500) #rldæ”¹ä¸ºvst
mat <- assay(rld)[top_var_genes, ]
mat <- t(scale(t(mat)))
pdf("/mnt/alamo01/users/chenyun730/program/test/results/Heatmap_geo.pdf", 
    width = 8, height = 10)
pheatmap(mat, 
         annotation_col = col_data,
         show_colnames = TRUE,
         show_rownames = FALSE,
         fontsize_row = 6ï¼Œ
          main = "Top 500 Genes by Variance of geo_matrix", )
dev.off()

```

3.å°†12ä¸ªæ ·æœ¬è¿›è¡Œåˆå¹¶åˆ†æ
**å‘ç°å‘½åé—®é¢˜å¹¶çŸ«æ­£â€”â€”symbol**
```
count_matrix <- read.csv("/mnt/alamo01/users/chenyun730/program/test/my_matrix/count_matrix.csv", row.names = NULL, check.names = FALSE)
count_matrix$symbol <- ifelse(grepl("\\|", count_matrix$gene_id),
                              sub(".*\\|", "", count_matrix$gene_id),
                              count_matrix$gene_id)
 library(dplyr)
 symbol_matrix <- count_matrix %>%
  select(-gene_id) %>%
  group_by(symbol) %>%
  summarise(across(everything(), sum)) %>%
  ungroup()
symbol_matrix <- symbol_matrix %>%
  relocate(symbol, .before = everything())
 write.csv(symbol_matrix, "/mnt/alamo01/users/chenyun730/program/test/my_matrix/symbol_count_matrix.csv", row.names = FALSE)
```
 è·å¾—åŸä½œè€…çš„çŸ©é˜µcount_matrix_geo.csv
 ```
 counts_all <- read.table("/mnt/alamo01/users/chenyun730/program/test/geo_matrix/GSE255647_all.counts.tsv.gz",header = TRUE, sep = "\t", row.names = 1, check.names = FALSE)
selected_samples <- c("MOCK_24hpi_rep1", 
                      "MOCK_24hpi_rep2", 
                      "MOCK_24hpi_rep3", 
                      "SARS-2_24hpi_rep1", 
                      "SARS-2_24hpi_rep2", 
                      "SARS-2_24hpi_rep3")
count_matrix_geo <- counts_all[, selected_samples]
count_matrix_geo <- cbind(gene_id = rownames(count_matrix_geo), count_matrix_geo)
colnames(count_matrix_geo)[2:7] <- c("Mock-1", "Mock-2", "Mock-3", 
                                     "SARS-2-1", "SARS-2-2", "SARS-2-3")
write.csv(count_matrix_geo,
          file = "/mnt/alamo01/users/chenyun730/program/test/geo_matrix/count_matrix_geo.csv",
          row.names = FALSE,
          quote = FALSE)
```
 æ„å»ºåˆå¹¶çŸ©é˜µï¼ˆmerged_counts)
```
df1 <- read.csv("/mnt/alamo01/users/chenyun730/program/test/my_matrix/symbol_count_matrix.csv", check.names = FALSE)
df2 <- read.csv("/mnt/alamo01/users/chenyun730/program/test/geo_matrix/symbol_count_matrix_geo.csv", check.names = FALSE)
colnames(df2)[2:ncol(df2)] <- paste0("geo.", colnames(df2)[2:ncol(df2)])
 merged_df <- full_join(df1, df2, by = "symbol")
 merged_df[is.na(merged_df)] <- 0
 write.csv(merged_df, "/mnt/alamo01/users/chenyun730/program/test/results/merged_count_matrix.csv", row.names = FALSE)
```
PCAå›¾
```
library(ggplot2)
library(pheatmap)
library(dplyr)
library(tibble)
library(FactoMineR)
library(factoextra)
data <- read.csv("/mnt/alamo01/users/chenyun730/program/test/results/merged_count_matrix.csv", check.names = FALSE)
rownames(data) <- data$symbol
data <- data[, -1]
log_data <- log2(data + 1)
log_data_filtered <- log_data[apply(log_data, 1, function(x) sd(x) > 0), ]
pca_data <- t(log_data_filtered)
pca_result <- prcomp(pca_data, scale. = TRUE)
 group <- rownames(pca_data)
group <- ifelse(grepl("^geo\\.Mock", group), "geo.Mock",
         ifelse(grepl("^geo\\.SARS2", group), "geo.SARS2",
         ifelse(grepl("^Mock", group), "Mock", "SARS2")))
group <- factor(group, levels = c("Mock", "SARS2", "geo.Mock", "geo.SARS2"))
 pca_plot <- fviz_pca_ind(pca_result,
                         geom.ind = "point",
                         pointshape = 21,
                         col.ind = group,
                         palette = c("#1f77b4", "#d62728", "#2ca02c", "#9467bd"),
                         addEllipses = TRUE,
                         legend.title = "Group") +
  theme_minimal() +
  ggtitle("PCA of RNA-seq Samples")
ggsave("PCA_plot.png", plot = pca_plot, width = 6, height = 5, dpi = 300)
```
çƒ­å›¾
```
 library(RColorBrewer)
> data <- read.csv("/mnt/alamo01/users/chenyun730/program/test/results/merged_count_matrix.csv", header = TRUE, row.names = 1)
> count_data <- data[, 1:12]
sum(is.na(scaled_data)) #å»é™¤naå’Œå…¨æ˜¯0çš„
 scaled_data[is.na(scaled_data)] <- 0
scaled_data[is.infinite(scaled_data)] <- 0
nonzero_rows <- rowSums(count_data) > 0
filtered_data <- scaled_data[nonzero_rows, ]
sum(is.infinite(scaled_data))
 log_data <- log2(count_data + 1)
> scaled_data <- t(scale(t(log_data)))
> sample_groups <- c(rep("Mock", 3), rep("SARS2", 3), rep("geo.Mock", 3), rep("geo.SARS2", 3))
> annotation_col <- data.frame(Group = factor(sample_groups))
rownames(annotation_col) <- colnames(scaled_data)
> group_colors <- list(
  Group = c(
    Mock = "#1F77B4",
    SARS2 = "#FF7F0E",
    geo.Mock = "#2CA02C",
    geo.SARS2 = "#D62728"
  )
)
pheatmap(
  scaled_data,
  main = "Heatmap of merged_matrix",
  color = colorRampPalette(rev(brewer.pal(11, "RdBu")))(100),
  cluster_rows = TRUE,
  cluster_cols = TRUE,
  show_rownames = FALSE,
  show_colnames = TRUE,
  annotation_col = annotation_col,
  annotation_colors = group_colors,
  scale = "row",
  border_color = NA,
  fontsize_col = 12,
  angle_col = 45,
  filename = "/mnt/alamo01/users/chenyun730/program/test/results/heatmap_merged_matrix.pdf",
  width = 10,
  height = 14
)
```
**å°†outcomeå’Œgeoè¿›è¡Œåˆ†ç»„è¿›è¡Œå·®å¼‚åˆ†æ**
```
 library(DESeq2)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggplot2)
library(pheatmap)
count_data <- read.csv("/mnt/alamo01/users/chenyun730/program/test/results/merged_count_matrix.csv", row.names = 1)
 condition <- factor(c(rep("outcome", 6), rep("geo", 6)))
 coldata <- data.frame(condition)
rownames(coldata) <- colnames(count_data)
 stopifnot(identical(rownames(coldata), colnames(count_data)))
 dds <- DESeqDataSetFromMatrix(countData = count_data,
                              colData = coldata,
                              design = ~ condition)
 dds <- dds[rowSums(counts(dds)) > 10, ]
 dds <- DESeq(dds)
 res <- results(dds, contrast = c("condition", "geo", "outcome"))
 res <- res[order(res$padj), ]
 write.csv(as.data.frame(res), file = "DESeq2_diff_expr_results.csv")
```

**æ€è€ƒé¢˜**ï¼š

å¦‚æœPCAä¸­æ ·æœ¬æœªèƒ½æŒ‰ç»„åˆ†ç¦»ï¼ŒåŸå› å¯èƒ½æ˜¯ä»€ä¹ˆï¼Ÿ

æ•°æ®ä¸ºæ ‡å‡†åŒ–ï¼ˆè¿™é‡Œåšäº†ï¼‰ï¼šlog2 è½¬æ¢ä»¥åŠscale.=TRUEï¼›å¤ªå¤šé›¶è¡¨è¾¾æˆ–ä½å˜å¼‚çš„åŸºå› ï¼Œå¯èƒ½æ©ç›–äº†çœŸå®çš„ç»„é—´å·®å¼‚ï¼›ç»„é—´å·®å¼‚å¤ªå°å¯¼è‡´çš„ç”Ÿç‰©å·®å¼‚ä¸æ˜æ˜¾ç­‰åŸå› ã€‚

### æ¨¡å—5ï¼šå·®å¼‚è¡¨è¾¾åˆ†æä¸åŠŸèƒ½å¯Œé›†
**ç›®æ ‡**ï¼šå·®å¼‚è¡¨è¾¾åˆ†æä¸åŠŸèƒ½å¯Œé›†ï¼Œç†è§£ç”Ÿç‰©å­¦æ„ä¹‰ã€‚

**ä»»åŠ¡**ï¼š

DESeq2ç­›é€‰å·®å¼‚è¡¨è¾¾åŸºå› (outcomeä¸geoï¼‰
```
# outcome_matrix
count_data <- read.csv("/mnt/alamo01/users/chenyun730/program/test/my_matrix/symbol_count_matrix.csv",row.names = 1)
 condition <- factor(c(rep("mock", 3), rep("sars2", 3)))
count_data[is.na(count_data)] <- 0
 coldata <- data.frame(condition)
rownames(coldata) <- colnames(count_data)
 stopifnot(identical(rownames(coldata), colnames(count_data)))
 dds <- DESeqDataSetFromMatrix(countData = count_data,
                              colData = coldata,
                              design = ~ condition)
 dds <- dds[rowSums(counts(dds)) > 10, ]
 dds <- DESeq(dds)
 res <- results(dds, contrast = c("condition","mock","sars2"))
 res <- res[order(res$padj), ]
 write.csv(as.data.frame(res), file = "/mnt/alamo01/users/chenyun730/program/test/my_matrix/outcome_DESeq2_diff_expr_results.csv")

# geo_matrix
count_data <- read.csv("/mnt/alamo01/users/chenyun730/program/test/geo_matrix/count_matrix_geo.csv",row.names = 1)
 condition <- factor(c(rep("mock", 3), rep("sars2", 3)))
count_data[is.na(count_data)] <- 0
 coldata <- data.frame(condition)
rownames(coldata) <- colnames(count_data)
 stopifnot(identical(rownames(coldata), colnames(count_data)))
 dds <- DESeqDataSetFromMatrix(countData = count_data,
                              colData = coldata,
                              design = ~ condition)
 dds <- dds[rowSums(counts(dds)) > 10, ]
 dds <- DESeq(dds)
 res <- results(dds, contrast = c("condition","mock","sars2"))
 res <- res[order(res$padj), ]
 write.csv(as.data.frame(res), file = "/mnt/alamo01/users/chenyun730/program/test/geo_matrix/geo_DESeq2_diff_expr_results.csv")

```
**clusterProfilerè¿›è¡ŒGO/KEGGå¯Œé›†åˆ†æ**

```
library(clusterProfiler)
library(org.Hs.eg.db) 
library(enrichplot)
library(DOSE)

# ç­›é€‰æ˜¾è‘—å·®å¼‚åŸºå› 
deg <- as.data.frame(res)
deg <- na.omit(deg)
deg_filtered <- deg[deg$padj < 0.05 & abs(deg$log2FoldChange) > 1, ]

# æå– gene symbol æˆ– ENSEMBL id
gene_symbols <- rownames(deg_filtered)
# symbolæˆ– "ENSEMBL"ï¼Œæ ¹æ®gene ID ç±»å‹ è½¬ Entrez ID
gene_entrez <- bitr(gene_symbols,
                    fromType = "SYMBOL", 
                    toType = "ENTREZID",
                    OrgDb = org.Hs.eg.db)

# æå– ENTREZID å‘é‡
entrez_ids <- gene_entrez$ENTREZID
ego <- enrichGO(gene         = entrez_ids,
                OrgDb        = org.Hs.eg.db,
                keyType      = "ENTREZID",
                ont          = "ALL",       # BP/CC/MF/ALL
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.2)

# ä¿å­˜ç»“æœ
write.csv(as.data.frame(ego), file = "GO_enrichment_results.csv")

# å¯è§†åŒ–ï¼ˆæŸ±çŠ¶å›¾ / ç‚¹å›¾ï¼‰
png("GO_barplot_geo.png", width = 800, height = 600)
barplot(ego, showCategory = 20, title = "GO Enrichment_geo")
dev.off()
png("GO_dotplot_geo.png", width = 800, height = 600)
dotplot(ego, showCategory = 20, title = "GO Enrichment_geo")
dev.off()
```
**GSEA**
```
library(clusterProfiler)
library(org.Hs.eg.db)
library(DOSE)  # å¦‚æœåˆ†æ KEGG å¯ä»¥ç”¨åˆ°
library(tibble)
library(enrichplot)
# å·²è¿è¡ŒDEseq2å¾—åˆ°res
# å»æ‰ NA
res <- na.omit(res)

# æå– SYMBOL å’Œ log2FoldChangeï¼Œè½¬ä¸º Entrez ID
res$symbol <- rownames(res)
gene_info <- bitr(res$symbol,
                  fromType = "SYMBOL",
                  toType = "ENTREZID",
                  OrgDb = org.Hs.eg.db)

# åˆå¹¶
res_merged <- merge(res, gene_info, by.x = "symbol", by.y = "SYMBOL")

# æ„é€ æ’åºå‘é‡ï¼šENTREZID ä¸º namesï¼Œlog2FC ä¸ºå€¼
gene_list <- res_merged$log2FoldChange
names(gene_list) <- res_merged$ENTREZID

# æŒ‰ log2FC é™åºæ’åºï¼ˆéå¸¸é‡è¦ï¼ï¼‰
gene_list <- sort(gene_list, decreasing = TRUE)

# è¿è¡Œ GSEAï¼ˆä»¥ GO ä¸ºä¾‹ï¼‰
ego_gsea <- gseGO(geneList = gene_list,
                  OrgDb = org.Hs.eg.db,
                  ont = "BP",
                  keyType = "ENTREZID",
                  nPerm = 1000,
                  minGSSize = 10,
                  maxGSSize = 500,
                  pvalueCutoff = 0.05,
                  verbose = TRUE)

# ä¿å­˜å¯Œé›†å›¾
png("GSEA_dotplot.png", width = 1000, height = 800)
dotplot(ego_gsea, showCategory = 20, title = "GSEA-GO BP")
dev.off()

# ä¹Ÿå¯ä»¥ç”» GSEA å•æ¡è·¯å¾„å›¾ï¼ˆæ¯”å¦‚ top1ï¼‰
png("GSEA_pathway1.png", width = 1000, height = 600)
gseaplot2(ego_gsea, geneSetID = ego_gsea$ID[1], title = ego_gsea$Description[1])
dev.off()

```
è¾“å‡ºï¼šå·®å¼‚è¡¨è¾¾ç»“æœã€å¯Œé›†ç»“æœ

æ€è€ƒé¢˜ï¼š

ä¸ºä»€ä¹ˆè¦è¿›è¡Œå¤šé‡æ£€éªŒæ ¡æ­£ï¼Ÿ

### æ¨¡å—å…­ æ›´æ¢æ¯”å¯¹æ–¹æ³•éªŒè¯è·Ÿgeoçš„å·®å¼‚
**1. ä¸‹è½½STARæ–‡ä»¶å¹¶æ„å»ºå‚è€ƒåŸºå› ç»„ï¼Œæ¯”å¯¹å®šé‡å¾—åˆ°countçŸ©é˜µstar_count_matrix.txt**
```
#æ–°å»ºä¸€ä¸ªtest_staræ–‡ä»¶å‚¨å­˜è¯¥æ¨¡å—çš„æ–‡ä»¶
#download.shç”¨äºä¸‹è½½STARçš„GRCh38ï¼ˆ2.7.0ï¼‰åŠæ³¨é‡Šæ–‡ä»¶
vim download.sh
#! /bin/bash
#micromamba activate R441
wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_42/GRCh38.primary_assembly.genome.fa.gz
gunzip GRCh38.primary_assembly.genome.fa.gz
wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_42/gencode.v42.primary_assembly.annotation.gtf.gz
gunzip gencode.v42.primary_assembly.annotation.gtf.gz

qsub -cwd -V -l cpu=64:mem=64G -q fast -N homodata_download /mnt/alamo01/users/chenyun730/program/test_star/scripts/download.sh

#star_index.shç”¨äºæ„å»ºç´¢å¼•
 micromamba create -n star_env -c bioconda -c conda-forge star=2.7.10a
 micromamba activate star_env
 STAR --version
vim star_index.sh
#! /bin/bash
#micromamba activate R441
module load STAR/2.7.10a
STAR --runThreadN 8 \
     --runMode genomeGenerate \
     --genomeDir star_index \
     --genomeFastaFiles GRCh38.primary_assembly.genome.fa \
     --sjdbGTFfile gencode.v42.primary_assembly.annotation.gtf \
     --sjdbOverhang 100

 qsub -cwd -V -l cpu=64:mem=64G -q fast -N star_index /mnt/alamo01/users/chenyun730/program/test_star/scripts/star_index.sh

#è¿›è¡Œæ¯”å¯¹
vim star_align.sh
#! /bin/bash
#micromamba activate R441
module load STAR/2.7.10a
INPUT_DIR=/mnt/alamo01/users/chenyun730/program/test_hisat2/clean_data
OUTPUT_DIR=/mnt/alamo01/users/chenyun730/program/test_star/alignment
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
    echo "Processing sample: ${SAMPLE}"
STAR --runThreadN 8 \
     --genomeDir /mnt/alamo01/users/chenyun730/program/test_star/homodata/star_index \
     --readFilesIn ${INPUT_DIR}/${SAMPLE}_cleaned_1.fp.gz ${INPUT_DIR}/${SAMPLE}_cleaned_2.fp.gz \
     --readFilesCommand zcat \
     --outFileNamePrefix ${OUTPUT_DIR}/${SAMPLE}/${SAMPLE}_ \
     --outSAMtype BAM SortedByCoordinate \
     --quantMode GeneCounts \
     --twopassMode Basic
done

# ä½¿ç”¨FeatureCountå®šé‡
vim  featurecounts.sh
#! /bin/bash
#micromamba activate R441
BAM_DIR=/mnt/alamo01/users/chenyun730/program/test_star/alignment
GTF_FILE=/mnt/alamo01/users/chenyun730/program/test_star/homodata/gencode.v42.primary_assembly.annotation.gtf
OUTPUT_DIR=/mnt/alamo01/users/chenyun730/program/test_star/quantification
COUNT_FILE=${OUTPUT_DIR}/gene_counts.txt
SUMMARY_FILE=${OUTPUT_DIR}/summary.txt
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
BAM_FILES=()
for SAMPLE in "${SAMPLES[@]}"; do
    BAM_FILES+=("${BAM_DIR}/${SAMPLE}/${SAMPLE}_Aligned.sortedByCoord.out.bam")
done
featureCounts \
    -T 8 \
    -a ${GTF_FILE} \
    -o ${COUNT_FILE} \
    -g gene_name \
    -p \
     --countReadPairs \
     -s 1 \
     --verbose \
    ${BAM_FILES[@]} > ${SUMMARY_FILE} 2>&1
cut -f1,7- ${COUNT_FILE} | grep -v '^#' | \
awk 'BEGIN{OFS="\t"}
     NR==1 {print "gene_id", "mock_1", "mock_2", "mock_3", "sars2_1", "sars2_2", "sars2_3"; next}
     {print $0}' > ${OUTPUT_DIR}/star_count_matrix.csv
```

**2.æ„å»ºåˆå¹¶çŸ©é˜µstar_merged.csv**
```
#å…ˆè¯»å…¥ç”Ÿæˆçš„star_count_matrix.txtè½¬æˆsymbolå†ä»¥.csvè¾“å‡ºä¸ºsymbol_star_count.csv
 df<- read.delim("/mnt/alamo01/users/chenyun730/program/test_star/quantification/star_count_matrix.txt", row.names = 1, check.names = FALSE)
 df$gene_id <- rownames(df)
rownames(df) <- NULL
 is_ensembl <- grepl("^ENSG", df$gene_id)
 ensembl_ids <- as.character(df$gene_id[is_ensembl])
df$symbol[is_ensembl] <- id_map$SYMBOL[match(df$gene_id[is_ensembl], id_map$ENSEMBL)]
df$symbol <- df$gene_id
df$symbol[is_ensembl] <- id_map$SYMBOL[match(df$gene_id[is_ensembl], id_map$ENSEMBL)]
library(dplyr)
df_clean <- df %>%
filter(!is.na(symbol) & symbol != "") %>%
select(-gene_id) %>%
group_by(symbol) %>%
summarise(across(everything(), sum)) %>%
ungroup() %>%
relocate(symbol)
write.csv(df_clean, "/mnt/alamo01/users/chenyun730/program/test_star/my_matrix/symbol_star_count.csv", row.names = FALSE)

# åˆå¹¶ä¸¤ä¸ªsymbolçŸ©é˜µä¸ºstar_merged.csv
df1<- read.csv("/mnt/alamo01/users/chenyun730/program/test_star/my_matrix/symbol_star_count.csv",row.names = 1, check.names = FALSE)
df2<- read.csv("/mnt/alamo01/users/chenyun730/program/test_hisat2/geo_matrix/symbol_count_matrix_geo.csv",row.names = 1, check.names = FALSE)
merged_df <- full_join(df1, df2, by = "symbol")
merged_df[is.na(merged_df)] <- 0
write.csv(merged_df, "/mnt/alamo01/users/chenyun730/program/test_star/results/star_merged.csv", row.names = FALSE)
```
PCA
```
library(ggplot2)
library(pheatmap)
library(dplyr)
library(tibble)
library(FactoMineR)
library(factoextra)
data <- read.csv("/mnt/alamo01/users/chenyun730/program/test_star/my_matrix/symbol_star_count.csv", check.names = FALSE)
rownames(data) <- data$symbol
data <- data[, -1]
log_data <- log2(data + 1)
log_data_filtered <- log_data[apply(log_data, 1, function(x) sd(x) > 0), ]
pca_data <- t(log_data_filtered)
pca_result <- prcomp(pca_data, scale. = TRUE)
 group <- rownames(pca_data)
group = factor(rep(c("mock", "sars2"), each = 3))
 pca_plot <- fviz_pca_ind(pca_result,
                         geom.ind = "point",
                         pointshape = 21,
                         col.ind = group,
                         palette = c("#1f77b4", "#d62728", "#2ca02c", "#9467bd"),
                         addEllipses = TRUE,
                         legend.title = "Group") +
  theme_minimal() +
  ggtitle("PCA of RNA-seq Samples")
ggsave("PCA_plot.png", plot = pca_plot, width = 6, height = 5, dpi = 300)
```

**ç”¨Stingtieä¹Ÿä¸€æ ·å¤„ç†STARæ¯”å¯¹åçš„bamæ–‡ä»¶ï¼Œç„¶åå¾—åˆ°stingtie_merged.csvæ•°æ®**
```
#ç”Ÿæˆ.gtfæ–‡ä»¶
vim stringtie_star.sh
#! /bin/bash
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
    echo "Processing sample ${SAMPLE}..."
     stringtie /mnt/alamo01/users/chenyun730/program/test_star/alignment/${SAMPLE}/${SAMPLE}_Aligned.sortedByCoord.out.bam \
        -G /mnt/alamo01/users/chenyun730/program/test_star/homodata/gencode.v42.primary_assembly.annotation.gtf \
        -o ./${SAMPLE}/${SAMPLE}.gtf  \
         -e \
         -B \
         -p 8
done
python3 /mnt/alamo01/users/chenyun730/program/test_hisat2/scripts/prepDE.py3 .

#ç”Ÿæˆåˆå¹¶çŸ©é˜µ
 df1<- read.csv("symbol_gene_count.csv", check.names = FALSE)  
 df2<- read.csv("/mnt/alamo01/users/chenyun730/program/test_hisat2/geo_matrix/symbol_count_matrix_geo.csv", check.names = FALSE)
 merged_df <- full_join(df1, df2, by = "symbol")                                 merged_df[is.na(merged_df)] <- 0
write.csv(merged_df, "/mnt/alamo01/users/chenyun730/program/test_star/results/stringtie_star_merged.csv", row.names = FALSE)

#PCA
data<- read.csv("/mnt/alamo01/users/chenyun730/program/test_star/results/stringtie_gene_star.csv",row.names = 1, check.names = FALSE)
rownames(data) <- data$symbol
data <- data[, -1]
log_data <- log2(data + 1)
log_data_filtered <- log_data[apply(log_data, 1, function(x) sd(x) > 0), ]
pca_data <- t(log_data_filtered)
pca_result <- prcomp(pca_data, scale. = TRUE)
 group <- rownames(pca_data)
group = factor(rep(c("mock", "sars2", each = 3))
 pca_plot <- fviz_pca_ind(pca_result,
                         geom.ind = "point",
                         pointshape = 21,
                         col.ind = group,
                         palette = c("#1f77b4", "#d62728", "#2ca02c", "#9467bd"),
                         addEllipses = TRUE,
                         legend.title = "Group") +
  theme_minimal() +
  ggtitle("PCA of RNA-seq Samples")
ggsave("/mnt/alamo01/users/chenyun730/program/test_star/results/PCA_stringtie_star.png", plot = pca_plot, width = 6, height = 5, dpi = 300)
Error: unexpected symbol in:
"group = factor(rep(c("mock", "sars2","geo_mock","geo_sars2", each = 3))
 pca_plot"

```
**æ£€æŸ¥å¤šæ˜ å°„readsæ¯”ä¾‹ï¼ˆSAM/BAMæ–‡ä»¶ä¸­çš„XSæ ‡ç­¾ï¼‰ï¼Œè¿‡é«˜ï¼ˆ>10%ï¼‰å¯èƒ½æç¤ºåŒæº/é‡å¤åºåˆ—é—®é¢˜**
```
$ samtools view -F 4 /mnt/alamo01/users/chenyun730/program/test_hisat2/alignment/SRR27961778/SRR27961778_sorted.bam | awk '{for (i=12; i<=NF; ++i) { if ($i ~ /^NH:i:/) { nh=substr($i,6); if (nh == 1) uniq++; else multi++; } } } END {print "Unique mappings:", uniq; print "Multi-mappings:", multi; print "Percentage multi-mapping:", multi/(uniq+multi)*100}'
Unique mappings: 1415394
Multi-mappings:
Percentage multi-mapping: 0
(R441) chenyun730@mgt01:/mnt/alamo01/users/chenyun730/program/test_hisat2
$ samtools view -F 4 /mnt/alamo01/users/chenyun730/program/test_hisat2/alignment/SRR27961779/SRR27961779_sorted.bam | awk '{for (i=12; i<=NF; ++i) { if ($i ~ /^NH:i:/) { nh=substr($i,6); if (nh == 1) uniq++; else multi++; } } } END {print "Unique mappings:", uniq; print "Multi-mappings:", multi; print "Percentage multi-mapping:", multi/(uniq+multi)*100}'
Unique mappings: 18106826
Multi-mappings:
Percentage multi-mapping: 0
(R441) chenyun730@mgt01:/mnt/alamo01/users/chenyun730/program/test_hisat2
$ samtools view -F 4 /mnt/alamo01/users/chenyun730/program/test_hisat2/alignment/SRR27961780/SRR27961780_sorted.bam | awk '{for (i=12; i<=NF; ++i) { if ($i ~ /^NH:i:/) { nh=substr($i,6); if (nh == 1) uniq++; else multi++; } } } END {print "Unique mappings:", uniq; print "Multi-mappings:", multi; print "Percentage multi-mapping:", multi/(uniq+multi)*100}'
Unique mappings: 19704911
Multi-mappings:
Percentage multi-mapping: 0
(R441) chenyun730@mgt01:/mnt/alamo01/users/chenyun730/program/test_hisat2
$ samtools view -F 4 /mnt/alamo01/users/chenyun730/program/test_hisat2/alignment/SRR27961787/SRR27961787_sorted.bam | awk '{for (i=12; i<=NF; ++i) { if ($i ~ /^NH:i:/) { nh=substr($i,6); if (nh == 1) uniq++; else multi++; } } } END {print "Unique mappings:", uniq; print "Multi-mappings:", multi; print "Percentage multi-mapping:", multi/(uniq+multi)*100}'
Unique mappings: 18680847
Multi-mappings:
Percentage multi-mapping: 0
(R441) chenyun730@mgt01:/mnt/alamo01/users/chenyun730/program/test_hisat2
$ samtools view -F 4 /mnt/alamo01/users/chenyun730/program/test_hisat2/alignment/SRR27961788/SRR27961788_sorted.bam | awk '{for (i=12; i<=NF; ++i) { if ($i ~ /^NH:i:/) { nh=substr($i,6); if (nh == 1) uniq++; else multi++; } } } END {print "Unique mappings:", uniq; print "Multi-mappings:", multi; print "Percentage multi-mapping:", multi/(uniq+multi)*100}'
Unique mappings: 18875024
Multi-mappings:
Percentage multi-mapping: 0
(R441) chenyun730@mgt01:/mnt/alamo01/users/chenyun730/program/test_hisat2
$ samtools view -F 4 /mnt/alamo01/users/chenyun730/program/test_hisat2/alignment/SRR27961789/SRR27961789_sorted.bam | awk '{for (i=12; i<=NF; ++i) { if ($i ~ /^NH:i:/) { nh=substr($i,6); if (nh == 1) uniq++; else multi++; } } } END {print "Unique mappings:", uniq; print "Multi-mappings:", multi; print "Percentage multi-mapping:", multi/(uniq+multi)*100}'
Unique mappings: 18252563
Multi-mappings:
Percentage multi-mapping: 0
(R441) chenyun730@mgt01:/mnt/alamo01/users/chenyun730/program/test_hisat2
$
```
