# GEO-RNAseq-Progress
Daily Updatesï¼šRecording of project progress

## ä»»åŠ¡èƒŒæ™¯ï¼š
åŸºäºæŒ‡å®šçš„GEOç¼–å·GSE255647ï¼Œé€šè¿‡RNA-seqæ•°æ®çš„ä¸‹è½½ã€é¢„å¤„ç†ã€æ¯”å¯¹ã€å®šé‡ã€å·®å¼‚è¡¨è¾¾åˆ†æå’ŒåŠŸèƒ½å¯Œé›†åˆ†æï¼Œå®Œæˆè½¬å½•ç»„æ•°æ®çš„å…¨æµç¨‹åˆ†æï¼Œå¹¶ä»¥Markdownå½¢å¼è®°å½•è¿‡ç¨‹ï¼Œæœ€ç»ˆä»¥GitHubä»“åº“å±•ç¤ºã€‚

**ä»»åŠ¡è¦æ±‚**ï¼š åˆ†æSARS-CoV-2ä»¥1 MOIæ„ŸæŸ“Calu-3/2B4ç»†èƒç³»12håï¼Œè½¬å½•ç»„æ°´å¹³çš„å˜åŒ–ã€‚

# âœ”å·²å®Œæˆ
## æ¨¡å—1ï¼šä»»åŠ¡å‡†å¤‡

**å®ç”¨æ“ä½œ**

lsï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šlist filesï¼‰: åˆ—å‡ºç›®å½•åŠæ–‡ä»¶å

cdï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šchange directoryï¼‰ï¼šåˆ‡æ¢ç›®å½•

pwdï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šprint work directoryï¼‰ï¼šæ˜¾ç¤ºç›®å‰çš„ç›®å½•

mkdirï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šmake directoryï¼‰ï¼šåˆ›å»ºä¸€ä¸ªæ–°çš„ç›®å½•

rmdirï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šremove directoryï¼‰ï¼šåˆ é™¤ä¸€ä¸ªç©ºçš„ç›®å½•

cpï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šcopy fileï¼‰: å¤åˆ¶æ–‡ä»¶æˆ–ç›®å½•

rmï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šremoveï¼‰: åˆ é™¤æ–‡ä»¶æˆ–ç›®å½•

mvï¼ˆè‹±æ–‡å…¨æ‹¼ï¼šmove fileï¼‰: ç§»åŠ¨æ–‡ä»¶ä¸ç›®å½•ï¼Œæˆ–ä¿®æ”¹æ–‡ä»¶ä¸ç›®å½•çš„åç§°



**å®Œæˆ**ï¼š

åŸºäº`micromamba`é…ç½®ç¯å¢ƒï¼Œå®‰è£…å¹¶ç†Ÿæ‚‰ä»¥ä¸‹å·¥å…·ï¼š

1. fastqcã€multiqcã€fastpï¼šè´¨æ§ä¸æ¸…æ´—

2. hisat2ã€stringtieï¼šæ¯”å¯¹ä¸å®šé‡

3. samtoolsï¼šBAMæ–‡ä»¶æ“ä½œ

4. Rä¸DESeq2ã€clusterProfilerã€ggplot2

5. æ³¨å†ŒGitHubè´¦å·ï¼ŒæŒæ¡åŸºæœ¬æ“ä½œ

**æ€è€ƒé¢˜**ï¼š

1. å¦‚ä½•åœ¨æ²¡æœ‰ç®¡ç†å‘˜æƒé™ä¸‹å®‰è£…fastqcï¼Ÿ

   ç­”ï¼šæœ‰å¤šç§æ–¹æ³•ï¼ŒåŒ…æ‹¬micromambaã€æ‰‹åŠ¨ä¸‹è½½äºŒè¿›åˆ¶æ–‡ä»¶ä»¥åŠåœ¨Javaç¯å¢ƒä¸‹é€šè¿‡æºç ç¼–è¯‘ç­‰æ–¹å¼ã€‚è¿™é‡Œä½¿ç”¨micromambaã€biocondaï¼ˆæˆ–é•œåƒï¼‰ï¼š
```
   micromamba create -p ~/micromamba_envs/fastqc_env -c bioconda -c conda-forge fastqc
```
2. GitHubçš„READMEæ–‡ä»¶æœ‰ä»€ä¹ˆä½œç”¨ï¼Ÿ
   
   é¡¹ç›®çš„â€œé—¨é¢â€å’Œâ€œä½¿ç”¨è¯´æ˜ä¹¦â€ï¼Œåƒè¿™æ ·ğŸ‘€ã€‚


## æ¨¡å—2ï¼šæ•°æ®ä¸‹è½½ä¸é¢„å¤„ç†

**ç›®æ ‡**ï¼šä»GEOä¸‹è½½åŸå§‹æ•°æ®ï¼Œè¿›è¡Œè´¨æ§å’Œæ¸…æ´—ã€‚

### ä»»åŠ¡ï¼š

**1. ä½¿ç”¨wgetã€curlæˆ–asperaä¸‹è½½æŒ‡å®šGEOç¼–å·çš„FASTQæ•°æ®(ä»¥SRR27961778ä¸ºä¾‹ï¼‰**
```
#ä¸‹è½½SRAæ–‡ä»¶
    wget -O SRR27961778.sra \"https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR27961778/SRR27961778"
#è½¬æ¢ä¸ºFASTQæ–‡ä»¶
   fastq-dump --split-files SRR27961778.sra
#å¯ä»¥å­˜ä¸ºå‹ç¼©åŒ…
   gzip SRR27961779_1.fastq
```

**2. è´¨æ§ï¼šfastqcå’Œmultiq**
```
#ç”ŸæˆåŸå§‹æ•°æ®è´¨æ§æŠ¥å‘Š
   fastqc SRR27961778_1.fastq.gz SRR27961778_2.fastq.gz -o ./raw_fastqc_results 
   cd /mnt/alamo01/users/chenyun730/program/test/raw_fastqc_results
   multiqc ./SRR*
```

**3. æ¸…æ´—ï¼šfastpï¼ˆå»ä½è´¨ã€å»æ¥å¤´ï¼‰**
```
#å…ˆåœ¨testé‡Œæ–°å»ºä¸€ä¸ªæ–‡ä»¶å¤¹cleanâ€”â€”dataç”¨äºå­˜å‚¨æ¸…æ´—åçš„æ•°æ®
$ cd /mnt/alamo01/users/chenyun730/program/test
$ mkdir clean_data
$ for F in /mnt/alamo01/users/chenyun730/program/test/raw_data/*_1.fastq.gz; do    
 R=${F%_*}_2.fastq.gz;
 BASE=${F##*/}; SAMPLE=${BASE%_*};
 time fastp -i $F -I $R
   -o /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_1.fp.gz
   -O /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_2.fp.gz
   -e 25 -q 30 -u 10 -r -M 30 -W 5 -w 64 -f 10 -F 10
   --adapter_sequence AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC
   --adapter_sequence_r2 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT
   -h /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_report.html
   -j /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_report.json &
done
#e â€”â€”å¹³å‡è´¨é‡é˜ˆå€¼ï¼› q â€”â€”è´¨é‡é˜ˆå€¼ï¼ˆé»˜è®¤ä¸º15ï¼Œè¿™é‡Œæé«˜åˆ°30ä¸ºæ›´ä¸¥æ ¼ï¼‰ï¼›
u â€”â€”å…è®¸è¾¾æ ‡ç¢±åŸºçš„æ¯”ä¾‹ä¸º10%ï¼› r â€”â€”å¯ç”¨ polyG ä¿®å‰ªï¼ˆé€‚ç”¨äº NovaSeqï¼‰ï¼›
M â€”â€”polyG çš„æœ€å°é•¿åº¦é˜ˆå€¼ï¼›
W 5 â€”â€”æ»‘åŠ¨çª—å£å¤§å°ï¼ˆ5bpï¼‰ï¼› w 64 â€”â€”ä½¿ç”¨64ä¸ªçº¿ç¨‹ï¼›
f 10 F 10 â€”â€”åˆ‡é™¤ Read 1/2 å‰ 10 bp
```

**4. è¾“å‡ºï¼šè´¨æ§æŠ¥å‘Šã€æ¸…æ´—åæ•°æ®**
```
multiqc ./ -n "Cleaned_Data_QC"
```

**æ€è€ƒé¢˜**ï¼š

å¦‚æœè´¨æ§æŠ¥å‘Šä¸­ç¢±åŸºè´¨é‡åä½ï¼Œå¦‚ä½•è°ƒæ•´fastpçš„å‚æ•°ï¼Ÿ
 
1.å…¨å±€è°ƒæ•´ï¼š æé«˜è´¨é‡é˜ˆå€¼qã€é™ä½å…è®¸æœªè¾¾æ ‡ç¢±åŸºçš„æ¯”ä¾‹uã€æé«˜å¹³å‡è´¨é‡é˜ˆå€¼eï¼›
   
2.è‹¥5' æˆ– 3' ç«¯è´¨é‡å·®ï¼Œåºåˆ—å¼€å¤´æˆ–æœ«å°¾è´¨é‡æ˜æ˜¾ä¸‹é™ï¼šæé«˜é¦–ä½åˆ‡é™¤çš„bpæ•°ï¼ˆ-fã€-Fã€-tï¼‰ï¼›
   
3.è‹¥æ¥å¤´åºåˆ—æ®‹ç•™ï¼ˆMultiQC çš„ Adapter Content éƒ¨åˆ†æ˜¾ç¤ºè¶…æ ‡ï¼‰ï¼šè‡ªåŠ¨æ£€æµ‹æ¥å¤´ï¼ˆå³ä½¿å·²æŒ‡å®šåºåˆ—ï¼‰æˆ–è€…é¢å¤–åˆ‡é™¤5bpï¼ˆ-trim_front1 5);
   
4.è‹¥é«˜æ¯”ä¾‹çš„é‡å¤åºåˆ—æˆ–ä½å¤æ‚åº¦åºåˆ—ï¼šè¿‡æ»¤æ‰é•¿åº¦< 50bp çš„ readsï¼ˆé»˜è®¤ 15ï¼‰-l 50ï¼›ç§»é™¤ä½å¤æ‚åº¦åºåˆ—-low_complexity_filterï¼›
   
5.è‹¥å±€éƒ¨åŒºåŸŸè´¨é‡æ³¢åŠ¨å¤§ï¼šå‡å°æ»‘åŠ¨çª—å£å¤§å°-Wï¼›æ»‘åŠ¨çª—å£å¹³å‡è´¨é‡é˜ˆå€¼ä» 30 é™è‡³ 25ï¼ˆæ›´å®½æ¾ï¼‰-M
    

## æ¨¡å—3ï¼šæ¯”å¯¹ä¸å®šé‡

**ç›®æ ‡**ï¼šåŸºäºå‚è€ƒåŸºå› ç»„æ„å»ºç´¢å¼•å¹¶è¿›è¡Œæ¯”å¯¹ä¸å®šé‡ã€‚

### ä»»åŠ¡ï¼š

ä¸‹è½½ç‰©ç§å‚è€ƒåŸºå› ç»„å’ŒåŸºå› æ³¨é‡Šï¼ˆGTFæ–‡ä»¶ï¼‰
```
 #ä¸‹è½½å‚è€ƒåŸºå› ç»„çš„FASTQå’ŒGTFæ–‡ä»¶
 cd /mnt/alamo01/users/chenyun730/program/test/compare
 mkdir homo_sapiens
 wget ftp://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
 wget ftp://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh38.104.gtf.gz
 # è§£å‹æ–‡ä»¶
 gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
 gunzip Homo_sapiens.GRCh38.104.gtf.gz
```

ç”¨hisat2-buildæ„å»ºç´¢å¼•ï¼Œhisat2æ¯”å¯¹
```
# å‚è€ƒåŸºå› ç»„ GRCh38.fa æ„å»ºç´¢å¼•ï¼ˆæäº¤è„šæœ¬å¹¶æŠ•é€’ä»»åŠ¡ï¼‰
mkdir -p /mnt/alamo01/users/chenyun730/program/test/scripts
mkdir -p /mnt/alamo01/users/chenyun730/program/test/logs

#æ–°å»ºshellè„šæœ¬
vim hisat2_index.sh
#! /usr/bin/sh
#micromamba activate R441
cd /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data
hisat2-build -p 64 GRCh38.fa GRCh38
2> /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/home_data

#èµ‹äºˆè„šæœ¬æ‰§è¡Œæƒé™
chmod 777 hisat2_index.sh

#æäº¤ä»»åŠ¡(æ— éœ€æŒ‡å®šé˜Ÿåˆ—ï¼‰
qsub -cwd -V -l cpu=64:mem=64G -q fast -N hisat2_index /mnt/alamo01/users/chenyun730/program/test/scripts/hisat2_index.sh
# qstatæŸ¥çœ‹ä»»åŠ¡é˜Ÿåˆ—ï¼›qstat -f ID æŸ¥çœ‹ä»»åŠ¡ï¼› qdel åˆ é™¤ä»»åŠ¡

# æ£€æŸ¥ç´¢å¼•æ–‡ä»¶
ls -lh /mnt/alamo01/users/chenyun730/program/test/compare/homo_sapiens/genome_index/

# è¿›è¡Œæ¯”å¯¹
#! /bin/bash
#source /mnt/alamo01/users/chenyun730/bin/micromamba
#micromamba activate R441
cd /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/
hisat2 -k 1 -p 64 \
  -x /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/GRCh38 \
  -S SRR27961779.sam \
  --novel-splicesite-outfile SRR27961779_junction.bed \
  --no-unal --dta \
  --un-conc-gz SRR27961779_unmapped.fq.gz \
  -1 /mnt/alamo01/users/chenyun730/program/test/clean_data/SRR27961779_cleaned_1.fp.gz \
  -2 /mnt/alamo01/users/chenyun730/program/test/clean_data/SRR27961779_cleaned_2.fp.gz \
  2> SRR27961779.align.stats

#å¤šä¸ªæ–‡ä»¶è¿›è¡Œæ¯”å¯¹
#! /bin/bash
#source /mnt/alamo01/users/chenyun730/bin/micromamba
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
cd /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/
for SAMPLE in "${SAMPLES[@]}"; do
hisat2 -k 1 -p 64 \
  -x /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/GRCh38 \
  -S ${SAMPLE}.sam \
  --novel-splicesite-outfile ${SAMPLE}_junction.bed \
  --no-unal --dta \
  --un-conc-gz ${SAMPLE}_unmapped.fq.gz \
  -1 /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_1.fp.gz \
  -2 /mnt/alamo01/users/chenyun730/program/test/clean_data/${SAMPLE}_cleaned_2.fp.gz \
  2> ${SAMPLE}.align.stats
echo "Submitted alignment job for ${SAMPLE}"
done
```
ä½¿ç”¨samtoolså’Œstringtieè¿›è¡Œè½¬å½•æœ¬ç»„è£…å’ŒåŸºå› å®šé‡
```
#è½¬æ¢SAMä¸ºBAMå¹¶æ’åº
$ vim samtools.sh
#! /bin/bash
#source /mnt/alamo01/users/chenyun730/bin/micromamba
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
cd /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/
for SAMPLE in "${SAMPLES[@]}"; do
echo "Processing ${SAMPLE}..."
  samtools sort -@ 8 -o ${SAMPLE}_sorted.bam ${SAMPLE}.sam
  samtools index ${SAMPLE}_sorted.bam
  echo "Finished processing ${SAMPLE}"
done

#ä½¿ç”¨stringtieè¿›è¡Œè½¬å½•æœ¬ç»„è£…å’ŒåŸºå› å®šé‡
$ vim stringtie.sh
#! /bin/bash
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
        echo "Processing ${SAMPLE}..."
        stringtie /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/${SAMPLE}_sorted.bam \
  -G /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/Homo_sapiens.GRCh38.109.gtf \
  -o /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/${SAMPLE}.gtf \
  -p 64 \
  -B
        echo "Finished processing ${SAMPLE}"
done

 ls /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/*.gtf > gtf_list.txt
 stringtie --merge -G /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/homo_data/Homo_sapiens.GRCh38.109.gtf   -o merged.gtf   gtf_list.txt
 
 #ç”¨sh.æäº¤ï¼Œè®¡ç®—æ ·æœ¬çš„åŸºå› è¡¨è¾¾é‡ï¼ˆä»¥åˆå¹¶åçš„åŸºå› ç»„åšå‚è€ƒï¼‰
#! /bin/bash
#micromamba activate R441
SAMPLES=(SRR27961778 SRR27961779 SRR27961780 SRR27961787 SRR27961788 SRR27961789)
for SAMPLE in "${SAMPLES[@]}"; do
    echo "Processing ${SAMPLE}..."
    stringtie /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/${SAMPLE}_sorted.bam \
        -G /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/alignment/merged.gtf \
         -o /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/quantify/${SAMPLE}.gtf \
         -e \
         -B \
         -A /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/quantify/${SAMPLE}_gene_abundance.tab
     echo "Finished processing ${SAMPLE}"
done
#ä¸‹è½½prepDE.py3è¿è¡Œå¾—åˆ° .csvæ–‡ä»¶
$ python3 /mnt/alamo01/users/chenyun730/program/test/scripts/prepDE.py3 -i /mnt/alamo01/users/chenyun730/program/test/homo_sapiens/quantify/quantify_gtf_list.txt -v

```

è¾“å‡ºï¼šæ¯”å¯¹ç»“æœï¼ˆBAMï¼‰ã€å®šé‡çŸ©é˜µ


**æ€è€ƒé¢˜**ï¼š

ä¸ºä»€ä¹ˆéœ€è¦æ„å»ºå‚è€ƒåŸºå› ç»„ç´¢å¼•ï¼Ÿ

1. åŠ é€Ÿå¯¹æ¯”ï¼Œå‡å°‘å¯¹æ¯”æ—¶é—´ï¼›
   
2. å ç”¨æ›´å°‘çš„å†…å­˜ï¼›
   
3. æ”¯æŒå¤æ‚æ¯”å¯¹æ¨¡å¼ï¼š
     ï¼ˆaï¼‰å‰ªæ¥æ¯”å¯¹ï¼šRNA-seqæ•°æ®éœ€æ£€æµ‹å¤–æ˜¾å­è¿æ¥ï¼Œç´¢å¼•ä¼šé¢„å­˜å‰ªåˆ‡ä½ç‚¹ä¿¡æ¯ï¼ˆHISAT2çš„snp_tranç´¢å¼•ï¼‰ï¼›
     ï¼ˆbï¼‰çªå˜å®¹å¿ï¼šåŒ…å«SNP/çªå˜çš„ç´¢å¼•ï¼ˆå¦‚genome_snp_tranï¼‰å¯æé«˜å¤šæ€æ€§æ ·æœ¬çš„æ¯”å¯¹ç‡ã€‚


# ğŸ¤·â€â™€ï¸æœªå®Œæˆ  
### æ¨¡å—4ï¼šæ•°æ®åˆ†æä¸å¯è§†åŒ–

**ç›®æ ‡**ï¼šå¯¹å®šé‡æ•°æ®è¿›è¡Œè´¨æ§å¹¶ä¸GEOåŸä½œè€…çš„è¡¨è¾¾çŸ©é˜µæ¯”è¾ƒã€‚

**ä»»åŠ¡**ï¼š

1. è¡¨è¾¾çŸ©é˜µçš„log2è½¬æ¢å’Œæ ‡å‡†åŒ–
```
#åŠ è½½RåŒ…å’Œ.csvæ•°æ®
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(FactoMineR)
library(factoextra)
>gene_counts <-read.csv("/mnt/alamo01/users/chenyun730/program/test/homo_sapiens/quantify/gene_count_matrix.csv", row.names = 1, header = TRUE)
>transcript_counts <- read.csv("/mnt/alamo01/users/chenyun730/program/test/homo_sapiens/quantify/transcript_count_matrix.csv", row.names = 1, header = TRUE)
>head(gene_counts)
>head(transcript_counts)

# log2 è½¬æ¢ï¼ˆåŠ  1 é¿å… log(0)ï¼‰
log_gene_counts <- log2(gene_counts + 1)
log_transcript_counts <- log2(transcript_counts + 1)

# Z-score å½’ä¸€åŒ–
zscore_gene_counts <- t(scale(t(log_gene_counts)))
zscore_transcript_counts <- t(scale(t(log_transcript_counts)))

# æ£€æŸ¥æ ‡å‡†åŒ–ç»“æœ
summary(zscore_gene_counts)
summary(zscore_transcript_counts)
```

2. çƒ­å›¾
```
# åŸºå› æ°´å¹³çƒ­å›¾
 # ä»…ä¿ç•™è‡³å°‘ 1 ä¸ªé NA å€¼çš„åŸºå› 
filtered_counts <- zscore_gene_counts[rowSums(is.na(zscore_gene_counts)) == 0, ]
filtered_counts <- filtered_counts[rowSums(is.nan(filtered_counts)) == 0, ]
 pheatmap(filtered_counts,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         scale = "row",
         show_rownames = FALSE,
         show_colnames = TRUE,
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "Gene Expression Heatmap ")

pheatmap(zscore_gene_counts,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         scale = "row",
         show_rownames = FALSE,
         show_colnames = TRUE,
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "Gene Expression Heatmap")

#è½¬å½•æœ¬æ°´å¹³çƒ­å›¾
  # ä»…ä¿ç•™è‡³å°‘ 1 ä¸ªé NA å€¼çš„åŸºå› 
filtered_counts_transcript <- zscore_transcript_counts[rowSums(is.na(zscore_transcript_counts)) == 0, ]
filtered_counts_transcript <- filtered_counts_transcript[rowSums(is.nan(filtered_counts_transcript)) == 0, ]

pheatmap(filtered_counts_transcript, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         scale = "row",
         show_rownames = FALSE,
         show_colnames = TRUE,
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "Transcript Expression Heatmap")
```
3.PCAåˆ†æ
```
# è®¡ç®— PCA
gene_pca_res <- PCA(t(zscore_gene_counts), graph = FALSE)

# PCA å¯è§†åŒ–
fviz_pca_ind(gene_pca_res,
             col.ind = "cos2", # æ ¹æ® cos2 è´¡çŒ®åº¦ç€è‰²
             gradient.cols = c("blue", "yellow", "red"),
             repel = TRUE, # é¿å…æ ‡ç­¾é‡å 
             title = "PCA of Gene Expression")
# è®¡ç®— PCA
transcript_pca_res <- PCA(t(zscore_transcript_counts), graph = FALSE)

# PCA å¯è§†åŒ–
fviz_pca_ind(transcript_pca_res,
             col.ind = "cos2",
             gradient.cols = c("blue", "yellow", "red"),
             repel = TRUE,
             title = "PCA of Transcript Expression")


```
4. è¾“å‡ºï¼šå¯è§†åŒ–ç»“æœï¼ˆçƒ­å›¾ã€PCAå›¾ï¼‰
```
# ä¿å­˜åŸºå› æ°´å¹³çƒ­å›¾
pdf("gene_heatmap.pdf", width = 8, height = 6)
pheatmap(zscore_gene_counts, cluster_rows = TRUE, cluster_cols = TRUE, scale = "row",
         show_rownames = FALSE, show_colnames = TRUE,
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "Gene Expression Heatmap")
dev.off()

# ä¿å­˜è½¬å½•æœ¬æ°´å¹³çƒ­å›¾
pdf("transcript_heatmap.pdf", width = 8, height = 6)
pheatmap(zscore_transcript_counts, cluster_rows = TRUE, cluster_cols = TRUE, scale = "row",
         show_rownames = FALSE, show_colnames = TRUE,
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "Transcript Expression Heatmap")
dev.off()

# ä¿å­˜åŸºå› æ°´å¹³ PCA å›¾
pdf("gene_PCA.pdf", width = 8, height = 6)
fviz_pca_ind(gene_pca_res, col.ind = "cos2",
             gradient.cols = c("blue", "yellow", "red"),
             repel = TRUE, title = "PCA of Gene Expression")
dev.off()

# ä¿å­˜è½¬å½•æœ¬æ°´å¹³ PCA å›¾
pdf("transcript_PCA.pdf", width = 8, height = 6)
fviz_pca_ind(transcript_pca_res, col.ind = "cos2",
             gradient.cols = c("blue", "yellow", "red"),
             repel = TRUE, title = "PCA of Transcript Expression")
dev.off()

```

**æ€è€ƒé¢˜**ï¼š

å¦‚æœPCAä¸­æ ·æœ¬æœªèƒ½æŒ‰ç»„åˆ†ç¦»ï¼ŒåŸå› å¯èƒ½æ˜¯ä»€ä¹ˆï¼Ÿ
